doc/type_erasure.md

```diff
@@ -0,0 +1,590 @@
+# Type Erasure in libunifex
+
+Type erasure is an important tool for managing compile-times by insulating different pieces of code from
+each other.  and also for implementing polymorphism
+
+The libunifex library provides several generic type-erasing wrappers built on top of `tag_invoke`-based customisation point objects (CPOs).
+These generic wrappers can be parameterised on a variadic list of CPOs that define the operations that concrete types must
+support and that the type-erased wrapper itself will expose.
+
+## Categories of type-erasing wrappers
+
+There are several different axes on which different type-erasing wrappers can make different decisions:
+* **Ownership** - Does the wrapper own the wrapped concrete object?
+  * **Unique Ownership** - The wrapper exclusively owns the lifetime of the concrete object it's wrapping.
+  * **Shared Ownership** - The wrapper has shared ownership of the concrete object it's wrapping.
+  * **No Ownership** - The lifetime of the wrapped object is managed outside of the wrapper.
+* **Object Storage** - How does the wrapper store the wrapped object?
+  * **Heap Allocated** - Wrapper stores the concrete object on the heap (possibly allocated using an allocator).
+  * **Stored In-Place** - Wrapper stores the concrete object in-place, inline within the storage of the wrapper object.
+  * **Hybrid Storage** - Wrapper stores small objects inline in the wrapper and larger objects on the heap. Commonly referred to as small-object optimisation.
+  * **External Storage** - Wrapper stores a reference/pointer to the object which is stored in memory not managed by the wrapper.
+* **VTable Storage** - How and where is the VTable for the type-erased object stored?
+  * **Indirect VTable** - Wrapper stores a pointer to the statically allocated vtable. Pointer indirection to call the method.
+  * **In-Place VTable** - Wrapper stores vtable entries in-line inside the wrapper.
+    Avoids pointer-indirection at cost of larger wrapper types. Also allows casting other type-erased types to wrapper types with narrower types.
```

Comment:
The mention of casting here was intended to refer to things like casting `any_object<cpo1, cpo2, cpo3>`  to `any_object<cpo1, cpo3>`
This cannot in general be supported using indirect vtables because the layout/ordering of the vtable entries is not necessarily compatible.

Reply:
Right. That's true. I was talking about whether we want to support "base" and "derived" type-erasing wrappers, where the derived type adds its entries to the vtable after the base entries. `folly::Poly` supports that, and automatically allows derived-to-base conversions.

Reply:
This is not a use-case I've considered. I'll think about how to address this but will leave it for another PR.